core/domain/email_manager.py

```diff
@@ -255,3 +297,57 @@ def send_moderator_action_email(
     _send_email(
         recipient_id, sender_id, intent, email_subject, full_email_content,
         bcc_admin=True)
+
+def send_exploration_membership_email(
+        sender_id, recipient_id, recipient_role, exploration_id,
+        exploration_title):
+    """Sends a email when a new user is given activity rights (Manager, Editor,
+    Viewer) to an exploration by creator of exploration.
+
+    Email will only be sent if recipient wants to receive these emails (i.e.
+    'can_receive_editor_role_email' is set True in recipent's preferences).
+    """
+    # Return from here if sending email is turned off.
+    if not feconf.CAN_SEND_EMAILS_TO_USERS:
+        log_new_error('This app cannot send emails to users.')
+        return
+
+    recipient_user_setting = user_services.get_user_settings(recipient_id)
+    sender_user_settings = user_services.get_user_settings(sender_id)
+    recipient_preferences = user_services.get_email_preferences(recipient_id)
+    if recipient_preferences['can_receive_editor_role_email'] is False:
+        log_new_error(
+            '%s will not receive email notification of being assigned to an '
+            'exploration' % recipient_user_setting.username)
+        return
+
+    if recipient_role == rights_manager.ROLE_OWNER:
```

Comment:
All this branching is a bit of a code smell. Better to have a dict keyed by the recipient role, whose values are dicts with the role/rights. Then just read the relevant value.

Also, I think the strings are misnamed. Better to call them role_string and rights_html, or something like that, to distinguish them from, e.g., the exploration rights object.